app/smc/pytest/e2e_stress.py

```diff
@@ -0,0 +1,275 @@
+PING_DMFW_DURATION_REG_ADDR = 0x80030448
+
+# ARC messages
+ARC_MSG_TYPE_TEST = 0x90
+ARC_MSG_TYPE_PING_DM = 0xC0
+ARC_MSG_TYPE_SET_WDT = 0xC1
+
+# Lower this number if testing local changes, so that tests run faster.
+MAX_TEST_ITERATIONS = 1000
+
+
+def get_arc_chip(unlaunched_dut: DeviceAdapter):
+    """
+    Validates the ARC firmware is alive and booted, since this required
+    for any test to run
+    """
+    # This is a hack- the RTT terminal doesn't work in pytest, so
+    # we directly call this internal API to flash the DUT.
+    unlaunched_dut.generate_command()
+    if unlaunched_dut.device_config.extra_test_args:
+        unlaunched_dut.command.extend(
+            unlaunched_dut.device_config.extra_test_args.split()
+        )
+    unlaunched_dut._flash_and_run()
+    time.sleep(1)
+    start = time.time()
+    # Attempt to detect the ARC chip for 15 seconds
+    timeout = 15
+    chips = []
+    while True:
+        try:
+            chips = pyluwen.detect_chips()
+        except Exception:
+            print("Warning- SMC firmware requires a reset. Rescanning PCIe bus")
+        if len(chips) > 0:
+            logger.info("Detected ARC chip")
+            break
+        time.sleep(0.5)
+        if time.time() - start > timeout:
+            raise RuntimeError("Did not detect ARC chip within timeout period")
+        rescan_pcie()
+    chip = chips[0]
+    try:
+        status = chip.axi_read32(ARC_STATUS)
+    except Exception:
+        print("Warning- SMC firmware requires a reset. Rescanning PCIe bus")
+        rescan_pcie()
+        status = chip.axi_read32(ARC_STATUS)
+    assert (status & 0xFFFF0000) == 0xC0DE0000, "SMC firmware postcode is invalid"
+    # Check post code status of firmware
+    assert (status & 0xFFFF) >= 0x1D, "SMC firmware boot failed"
+    return chip
+
+
+@pytest.fixture(scope="session")
+def arc_chip(unlaunched_dut: DeviceAdapter):
+    return get_arc_chip(unlaunched_dut)
+
+
+def tt_smi_reset():
+    """
+    Resets the SMC using tt-smi
+    """
+    smi_reset_cmd = "tt-smi -r"
+    smi_reset_result = subprocess.run(
+        smi_reset_cmd.split(), capture_output=True, check=False
+    ).returncode
+    return smi_reset_result
+
+
+def test_arc_watchdog(arc_chip):
+    """
+    Validates that the DMC firmware watchdog for the ARC will correctly
+    reset the chip
+    """
+    wdt_timeout = 1000
+    total_tries = min(MAX_TEST_ITERATIONS, 100)
+    for i in range(total_tries):
+        logger.info(f"Starting ARC watchdog test iteration {i}/{total_tries}")
+        # Setup ARC watchdog for a 1000ms timeout
+        arc_chip.arc_msg(ARC_MSG_TYPE_SET_WDT, True, False, wdt_timeout, 0, 1000)
+        # Sleep 1500, make sure we can still ping arc
+        time.sleep(1.5)
+        response = arc_chip.arc_msg(ARC_MSG_TYPE_TEST, True, False, 1, 0, 1000)
+        assert response[0] == 2, "SMC did not respond to test message"
+        assert response[1] == 0, "SMC response invalid"
```

Comment:
Mentioned a long time ago, but for stress tests, we don't want to fail the test if a single iteration doesn't work.

We want to record the failure and move on to the next iteration of the test. That way, if a test passes 999/1000 tries, or is above a specific threshold, then the test will still pass.

What should be asserted is whether the success rate is above a certain threshold, not that a single iteration failed.

Also, it would be great to record results for this so we can see it over time on a graph.

Twister has a feature for recording results, in tests/benchmarks

Reply:
I'll update this PR to record statistics, I agree that will be useful to have. However I think we should initially aim for 100% pass rate. This is how our current long tests are designed, and they pass nightly. If we are routinely failing a test than we can evaluate the impact of allowing a 99% or 99.9% pass rate